powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/ParamManager.java

```diff
@@ -51,11 +81,12 @@ public static SecretsProvider getSecretsProvider() {
      * @return a {@link SSMProvider}
      */
     public static SSMProvider getSsmProvider() {
-        if (ssmProvider == null) {
+        if (!providers.containsKey(SSMProvider.class)) {
             ssmProvider = SSMProvider.builder()
                             .withCacheManager(cacheManager)
                             .withTransformationManager(transformationManager)
                             .build();
+            providers.put(SSMProvider.class, ssmProvider);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **84 and 89**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **91**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.